File: src/ts/magma/api/collect/list/List.ts
import { Sequence } from "../../../../magma/api/collect/list/Sequence";
import { Option } from "../../../../magma/api/option/Option";
import { Iterable } from "../../../../magma/api/collect/list/Iterable";
export interface List<T> extends Sequence<T> {
	addLast(element: T): List<T>;
	subList(startInclusive: number, endExclusive: number): Option<List<T>>;
	addAll(others: Iterable<T>): List<T>;
	addFirst(element: T): List<T>;
	removeNode(element: T): List<T>;
	removeLast(): Option<List<T>>;
}

File: src/main/java/magma/list/List.java
package magma.list;

import magma.Tuple;
import magma.option.Option;

import java.util.Arrays;
import java.util.Comparator;

public interface List<T> {
	@SafeVarargs
	static <T> List<T> of(T... elements) {
		return new ArrayList<T>(new java.util.ArrayList<T>(Arrays.asList(elements)));
	}

	Stream<T> stream();

	List<T> addLast(T element);

	List<T> copy();

	List<T> addAll(List<T> others);

	int size();

	Option<T> get(int index);

	boolean isEmpty();

	Option<T> getLast();

	List<T> sort(Comparator<T> comparator);

	Option<Tuple<List<T>, T>> removeLast();

	Option<T> getFirst();

	List<T> subListOrEmpty(int start, int end);
}
